armory/instrument/instrument.py

```diff
@@ -0,0 +1,670 @@
+        """
+        count = 0
+        for probe_variable, filter_map in self.probe_filter_meter_arg.items():
+            for stage_filter, meters_args in filter_map.items():
+                count += len(meters_args)
+        return count
+
+    def __str__(self):
+        return f"{type(self)} : {self.probe_filter_meter_arg}"
+
+    def connect_meter(self, meter):
+        """
+        Connect meter to probes; idempotent
+        """
+        for arg in meter.get_arg_names():
+            probe_variable, stage_filter = process_meter_arg(arg)
+            if probe_variable not in self.probe_filter_meter_arg:
+                self.probe_filter_meter_arg[probe_variable] = {}
+            filter_map = self.probe_filter_meter_arg[probe_variable]
+            if stage_filter not in filter_map:
+                filter_map[stage_filter] = []
+            meters_args = filter_map[stage_filter]
+            if (meter, arg) in meters_args:
+                log.warning(
+                    f"(meter, arg) pair ({meter}, {arg}) already connected, not adding"
+                )
+            else:
+                meters_args.append((meter, arg))
+
+    def disconnect_meter(self, meter):
+        """
+        Disconnect meter from probes; idempotent
+        """
+        for arg in meter.get_arg_names():
+            probe_variable, stage_filter = process_meter_arg(arg)
+            if probe_variable not in self.probe_filter_meter_arg:
+                continue
+            filter_map = self.probe_filter_meter_arg[probe_variable]
+
+            if stage_filter not in filter_map:
+                continue
+            meters_args = filter_map[stage_filter]
+
+            if (meter, arg) in meters_args:
+                meters_args.remove((meter, arg))
+                if not meters_args:
+                    filter_map.pop(stage_filter)
+                if not filter_map:
+                    self.probe_filter_meter_arg.pop(probe_variable)
+
+    def map_to_meters_args(self, probe_variable, stage):
+        """
+        Return a list of (meter, arg) that are using the current probe_variable
+        """
+        filter_map = self.probe_filter_meter_arg.get(probe_variable, {})
+        meters = filter_map.get(stage, [])
+        meters.extend(filter_map.get(None, []))  # no stage filter (default)
+        return meters
+
+
+class Context:  # NOTE: may need to rename to something like Experiment or Procedure
+    def __init__(self, name="global"):
+        self.name = name
+        self.batch = -1
+        self.stage = ""
+        self.mapper = ProbeMapper()
+        self.meters = []
+        self.writers = []
+        self.closed = False
+
+    def set_stage(self, stage: str):
+        self.stage = stage
+
+    def set_batch(self, batch: int):
+        # NOTE: batch could be a set of sample indices
+        self.batch = batch
+
+    def increment_batch(self):
+        self.batch += 1
+
+    def is_measuring(self, probe_variable):
```

Comment:
I understand this may be by design, but I could imagine the following kind of thing causing confusion in the future:

if there's a misalignment between the string I pass to `add_meter()` and the `kwarg` passed to `probe.update()` nothing will happen. E.g. if I make a typo and pass

```
add_meter(
        "categorical_accuracy",
        metrics.categorical_accuracy,
        "scenaro.y",
        "scenario.y_pred",
```
categorical accuracy won't be measured, nor will I see any error or warning (since `is_measuring()` will return `False`). I'd need to think more about a way to catch these kinds of issues without violating design decisions about how a probe/meter/etc will work

Reply:
I like the idea of adding a warning, though not an error, if a value name is added to a meter but never captured. I think it's nice to enable conditional measurement (e.g., measuring targeted attacks but only when targeted attacks happen in the code) without erroring out.

 One challenge is that probes may not be connected when meters are added (and vice versa). In fact, a probe update that is hooked in model code will only happen during the forward pass.

I think a fix for this would be to raise a warning when all the meters are finalized if a value has never been read.

Reply:
Now the output will do something like:
```
2022-04-15 21:47:35  4s WARNING  __main__:finalize:452 Meter 'categorical_accuracy' was never measured. The following args were never set: ['scenaro.y']
2022-04-15 21:47:35  4s WARNING  __main__:finalize:452 Meter 'targeted_categorical_accuracy' was never measured. The following args were never set: ['scenario.y_target']
```

Reply:
I agree a warning makes the most sense. The new output looks good